subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,314 @@
+}
+
+#define EC_HOST_CMD_UART_DEFINE(_name)                                                             \
+	static struct ec_host_cmd_uart_ctx _name##_hc_uart;                                        \
+	static struct ec_host_cmd_backend _name = {                                                \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = &_name##_hc_uart,                                                           \
+	}
+
+/* Waiting time in microseconds to detect overrun */
+#define UART_OVERRUN_TIMEOUT_US 300
+/* Timeout after receiving first byte */
+#define UART_REQ_RX_TIMEOUT	K_MSEC(150)
+
+static void rx_timeout(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct ec_host_cmd_uart_ctx *hc_uart =
+		CONTAINER_OF(dwork, struct ec_host_cmd_uart_ctx, timeout_work);
+	int res;
+
+	switch (hc_uart->state) {
+	case UART_HOST_CMD_RECEIVING:
+		/* If state is receiving then timeout was hit due to underrun */
+		LOG_ERR("Request underrun detected");
+		break;
+	case UART_HOST_CMD_RX_OVERRUN:
+		/* If state is rx_overrun then timeout was hit because
+		 * process request was cancelled and extra rx bytes were
+		 * dropped
+		 */
+		LOG_ERR("Request overrun detected");
+		break;
+	case UART_HOST_CMD_RX_BAD:
+		/* If state is rx_bad then packet header was bad and process
+		 * request was cancelled to drop all incoming bytes.
+		 */
+		LOG_ERR("Bad packet header detected");
+		break;
+	default:
+		LOG_ERR("Request timeout mishandled, state: %d", hc_uart->state);
+	}
+
+	res = uart_rx_disable(hc_uart->uart_dev);
+	res = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER, UART_OVERRUN_TIMEOUT_US);
+
+	hc_uart->state = UART_HOST_CMD_READY_TO_RX;
+}
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct ec_host_cmd_uart_ctx *hc_uart = user_data;
+
+	switch (evt->type) {
+	case UART_RX_RDY:
+		size_t new_len;
+
+		if (hc_uart->state == UART_HOST_CMD_READY_TO_RX) {
+			hc_uart->rx_ctx->len = 0;
+			hc_uart->state = UART_HOST_CMD_RECEIVING;
+			k_work_reschedule(&hc_uart->timeout_work, UART_REQ_RX_TIMEOUT);
+		} else if (hc_uart->state == UART_HOST_CMD_PROCESSING ||
+			   hc_uart->state == UART_HOST_CMD_SENDING) {
+			LOG_ERR("UART HOST CMD ERROR: Received data while processing or sending");
+			return;
+		} else if (hc_uart->state == UART_HOST_CMD_RX_BAD ||
+			   hc_uart->state == UART_HOST_CMD_RX_OVERRUN) {
+			/* Wait for timeout if an error has been detected */
+			return;
+		}
+
+		__ASSERT(hc_uart->state != UART_HOST_CMD_RECEIVING,
+			 "UART Host Command state mishandled, state: %d", hc_uart->state);
+
+		new_len = hc_uart->rx_ctx->len + evt->data.rx.len;
+
+		if (new_len > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+			/* Bad data error, set the state and wait for timeout */
+			hc_uart->state = UART_HOST_CMD_RX_BAD;
+			return;
+		}
```

Comment:
I think `UART_HOST_CMD_RX_OVERRUN` is well defined - extra bytes are received after receiving the full command (number of bytes defined by header). `UART_HOST_CMD_RX_BAD` is more unclear. I think treating receiving more bytes than `CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER` is a good approach, because we send the protocol info HC, which includes max request size, so it is the host's fault.  The max request size is set based on the rx buffer size. At the end, the state is for internal debug/use only. In both cases we should hit timeout and reset communication. UART doesn't support reporting state.